core/src/main/java/org/elasticsearch/discovery/zen/ping/ZenPing.java

```diff
@@ -171,15 +193,15 @@ public synchronized boolean addPing(PingResponse ping) {
         }
 
         /** adds multiple pings if newer than previous pings from the same node */
-        public synchronized void addPings(PingResponse[] pings) {
+        public synchronized void addPings(Iterable<PingResponse> pings) {
             for (PingResponse ping : pings) {
                 addPing(ping);
             }
         }
 
-        /** serialize current pings to an array */
-        public synchronized PingResponse[] toArray() {
-            return pings.values().toArray(new PingResponse[pings.size()]);
+        /** serialize current pings to an array. It is guaranteed that the array contains one ping response per node */
+        public synchronized List<PingResponse> toList() {
+            return new ArrayList<>(pings.values());
```

Comment:
why do you copy it? can't you just use `Collections.unmodifiableCollection(ping.values())` and let the user do it if needed or do we have problems with concurrent modifications here? Maybe a `CopyOnWriteHashMap` would be the right thing todo?